src/endpoint.rs

```diff
@@ -149,11 +241,22 @@ impl Endpoint {
             },
             Err(e) => info!("Echo service timed out: {:?}", e),
         }
-        if let Some(socket_addr) = addr {
+
+        addr.map_or(Err(Error::NoEchoServiceResponse), |socket_addr| {
+            self.public_addr = Some(socket_addr);
             Ok(socket_addr)
+        })
+    }
+
+    /// Disconnects from a peer
+    pub fn disconnect_from(&mut self, peer_addr: &SocketAddr) -> Result<()> {
+        if let Some((conn, remover)) = self.connection_pool.remove(peer_addr) {
+            conn.close(0u8.into(), b"User Disconnected");
+            remover.remove();
```

Comment:
You already removed the connection from the pool by calling `ConnectionPool::remove` above, so there is no need to remove it via the remover too.